### DISCUSSION TITLE
How do you share a variable across multiple workers?

### DISCUSSION THREAD

**Discussion by yangh0597** [benoitc/gunicorn#3017] [Q&A]
How do you share a variable across multiple workers? Different pid

**Comment by zacqed** [comment]
We use redis as the caching layer for being able to share it across

**Reply by yangh0597** [reply]
This object is a complex object, hundreds of megabytes in size

**Comment by tilgovi** [comment]
Use an external store. Gunicorn workers are designed to share nothing.

**Comment by yangh0597** [comment]
In the field of machine learning, it is necessary to share a few hundred megabytes of objects, and then provide prediction functions through the interface, is there any good solution?

**Comment by tilgovi** [comment]
Again, Gunicorn workers are designed to share nothing. The only way to share things is to use OS methods of memory sharing.

Gunicorn uses `fork()` without `exec()`, so Gunicorn workers share any memory that was allocated before the worker started. Using the `preload` option or putting code in your config module, you _may_ be able to load Python modules and have (an initial snapshot of) their data shared between workers. Changes will _not_ be reflected across workers, though, because that's not how OS memory management works! Be aware that doing this can be brittle. In particular, macOS has recently become quite strict about what's allowed after `fork()` without `exec()`.

A common pattern for any web service that has heavy work is to submit that work to a queue and handle it with a separate service. You can have some lower memory machines serving web requests and higher memory machines performing the prediction work.

Consider whether you really need to run multiple prediction processes _on the same machine_. If you do, and you _really_ care about de-duplicating their memory consumption, look into ways to share memory between those processes. Sharing _Python objects_ is not easy. The `multiprocessing` module provides some ability to share `ctypes` objects, but not arbitrary Python objects. It can share arbitrary objects via proxies, but that's really just keeping the objects in one process. If you simply need to read a file, but are okay with building the Python objects in each process, consider just `mmap`.

**Comment by tilgovi** [comment]
Consider also whether maybe Python is not the best language for your memory-and-cpu-intensive work.
